File: src/serve/camera.js
// Soli camera preview — the client half.
//
// Injected only into pages carrying a `[data-soli-camera]` element, so a page
// without one pays nothing.
//
// What it does: starts a stream into every such <video>, keeps it alive across
// instant-nav swaps, and — the part hand-written code usually forgets — stops
// the tracks when the element goes away, so the camera indicator does not stay
// lit after the user has navigated on.
(function () {
  "use strict";

  if (window.__soliCameraStarted) return;
  window.__soliCameraStarted = true;

  var active = new WeakMap();

  function constraintsFor(element) {
    var facing = element.getAttribute("data-facing") || "user";
    var video = { facingMode: facing };

    var width = parseInt(element.getAttribute("data-width"), 10);
    var height = parseInt(element.getAttribute("data-height"), 10);
    // `ideal`, not `exact`: an exact constraint that no camera satisfies fails
    // the whole request rather than picking the nearest mode.
    if (width) video.width = { ideal: width };
    if (height) video.height = { ideal: height };

    return { video: video, audio: element.hasAttribute("data-audio") };
  }

  function fail(element, error) {
    element.setAttribute("data-camera-state", "error");
    // The page decides what to show; this only reports. `NotAllowedError` means
    // denied, `NotFoundError` means no camera, `NotReadableError` means another
    // app holds it.
    element.dispatchEvent(
      new CustomEvent("soli:camera-error", {
        bubbles: true,
        detail: { name: error && error.name, message: error && error.message }
      })
    );

    var fallback = element.getAttribute("data-fallback");
    if (fallback) {
      var node = document.querySelector(fallback);
      if (node) node.hidden = false;
    }
  }

  function start(element) {
    if (active.has(element)) return;
    if (!navigator.mediaDevices || !navigator.mediaDevices.getUserMedia) {
      fail(element, { name: "NotSupportedError", message: "getUserMedia is unavailable" });
      return;
    }

    element.setAttribute("data-camera-state", "starting");
    navigator.mediaDevices
      .getUserMedia(constraintsFor(element))
      .then(function (stream) {
        // The element may have been swapped out while the user was answering
        // the permission prompt; do not leave an orphan stream running.
        if (!element.isConnected) {
          stop(stream);
          return;
        }
        active.set(element, stream);
        element.srcObject = stream;
        element.setAttribute("data-camera-state", "live");
        element.dispatchEvent(new CustomEvent("soli:camera-ready", { bubbles: true }));
        if (element.hasAttribute("data-soli-scan")) startScanning(element);
        var playing = element.play();
        if (playing && playing.catch) playing.catch(function () {});
      })
      .catch(function (error) {
        fail(element, error);
      });
  }

  function stop(streamOrElement) {
    var stream = streamOrElement;
    if (streamOrElement instanceof Element) {
      stopScanning(streamOrElement);
      stream = active.get(streamOrElement);
      active.delete(streamOrElement);
      streamOrElement.srcObject = null;
      streamOrElement.setAttribute("data-camera-state", "stopped");
    }
    if (!stream || !stream.getTracks) return;
    stream.getTracks().forEach(function (track) {
      track.stop();
    });
  }

  /// A still frame as a data URL, for upload or preview.
  function snapshot(element, type, quality) {
    if (!element || !element.videoWidth) return null;
    var canvas = document.createElement("canvas");
    canvas.width = element.videoWidth;
    canvas.height = element.videoHeight;
    var context = canvas.getContext("2d");
    // A front-facing preview is mirrored for the user's benefit; the capture
    // should not be, or text in frame comes out backwards.
    if (element.getAttribute("data-facing") === "user" && !element.hasAttribute("data-no-unmirror")) {
      context.translate(canvas.width, 0);
      context.scale(-1, 1);
    }
    context.drawImage(element, 0, 0);
    return canvas.toDataURL(type || "image/jpeg", quality || 0.9);
  }

  // -------------------------------------------------------------------
  // Barcode / QR scanning
  // -------------------------------------------------------------------
  //
  // `BarcodeDetector` is native in Chromium — the Android shell, and Chrome on
  // Windows/Linux — and absent from WebKit, so the macOS shell and Safari need
  // a decoder supplied by the page. Soli deliberately ships the loop, not the
  // decoder: a WASM barcode reader is ~200 KB, and putting it in every soli
  // binary to serve the pages that scan would be the wrong trade.
  //
  //   window.soli.camera.decoder = async (video) => { ... return "text" | null }
  //
  // The loop, the throttling and the lifecycle are the fiddly parts, and those
  // are here.

  var scanners = new WeakMap();

  function detectorFor(formats) {
    if (typeof window.BarcodeDetector !== "function") return null;
    try {
      return new window.BarcodeDetector({ formats: formats });
    } catch (e) {
      // Thrown when none of the requested formats is supported.
      return null;
    }
  }

  function startScanning(element) {
    if (scanners.has(element)) return;

    var formats = (element.getAttribute("data-soli-scan") || "qr_code")
      .split(",")
      .map(function (f) { return f.trim(); })
      .filter(Boolean);
    var interval = parseInt(element.getAttribute("data-scan-interval"), 10) || 100;
    var continuous = element.hasAttribute("data-scan-continuous");

    var detector = detectorFor(formats);
    if (!detector && typeof (window.soli.camera || {}).decoder !== "function") {
      element.dispatchEvent(
        new CustomEvent("soli:scan-unsupported", {
          bubbles: true,
          detail: { formats: formats }
        })
      );
      return;
    }

    var state = { stopped: false };
    scanners.set(element, state);

    function found(value, format) {
      element.dispatchEvent(
        new CustomEvent("soli:scan", {
          bubbles: true,
          detail: { value: value, format: format || null }
        })
      );
      if (!continuous) stopScanning(element);
    }

    function tick() {
      if (state.stopped || !element.isConnected) return;
      // A video that is not playing yet has no frame to read.
      if (element.readyState < 2) {
        setTimeout(tick, interval);
        return;
      }

      var attempt = detector
        ? detector.detect(element).then(function (codes) {
            return codes && codes.length ? { value: codes[0].rawValue, format: codes[0].format } : null;
          })
        : Promise.resolve(window.soli.camera.decoder(element)).then(function (value) {
            return value ? { value: value, format: null } : null;
          });

      attempt
        .then(function (hit) {
          if (state.stopped) return;
          if (hit) found(hit.value, hit.format);
          if (!state.stopped) setTimeout(tick, interval);
        })
        .catch(function () {
          // A single bad frame is not a reason to stop scanning.
          if (!state.stopped) setTimeout(tick, interval);
        });
    }

    tick();
  }

  function stopScanning(element) {
    var state = scanners.get(element);
    if (state) state.stopped = true;
    scanners.delete(element);
  }

  function scan(root) {
    (root || document).querySelectorAll("[data-soli-camera]").forEach(function (element) {
      if (element.hasAttribute("data-manual")) return;
      start(element);
    });
  }

  // Release cameras whose element left the DOM — an instant-nav body swap does
  // not fire unload, so nothing else would.
  var observer = new MutationObserver(function (mutations) {
    mutations.forEach(function (mutation) {
      mutation.removedNodes.forEach(function (node) {
        if (node.nodeType !== 1) return;
        if (node.matches && node.matches("[data-soli-camera]")) stop(node);
        if (node.querySelectorAll) {
          node.querySelectorAll("[data-soli-camera]").forEach(stop);
        }
      });
    });
  });

  function boot() {
    scan(document);
    observer.observe(document.documentElement, { childList: true, subtree: true });
  }

  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", boot);
  } else {
    boot();
  }

  // Instant navigation replaces the body without a page load.
  document.addEventListener("soli:load", function () {
    scan(document);
  });

  window.addEventListener("pagehide", function () {
    document.querySelectorAll("[data-soli-camera]").forEach(stop);
  });

  window.soli = window.soli || {};
  window.soli.camera = {
    start: start,
    stop: stop,
    snapshot: snapshot,
    startScanning: startScanning,
    stopScanning: stopScanning,
    /// Whether a code can be read without the page supplying a decoder.
    scanningIsNative: typeof window.BarcodeDetector === "function",
    /// Whether this host can offer a camera at all, for deciding what to render.
    available: !!(navigator.mediaDevices && navigator.mediaDevices.getUserMedia)
  };
})();

File: src/serve/camera.rs
//! Camera preview and barcode scanning — server half.
//!
//! Ships `camera.js` at `/__soli/camera.js`, injected only into pages that
//! actually use a camera. The gate is the markup itself: a `data-soli-camera`
//! attribute, which `camera_preview(...)` emits.
//!
//! Showing a camera needs no framework at all — `getUserMedia` into a `<video>`
//! is six lines, and works as soon as the shell stops denying it. What the
//! script adds is the part hand-written code reliably forgets: **stopping the
//! tracks**. Instant navigation swaps the body without a page unload, so a
//! stream started by a page that has since been replaced keeps running, and the
//! camera indicator stays lit. That, plus a scan loop that is throttled rather
//! than running at frame rate, is the whole justification for its existence.

use bytes::Bytes;
use http_body_util::Full;
use hyper::Response;

use crate::serve::live_reload::rfind_ascii_case_insensitive;
use crate::serve::prefetch;

/// Client JS — compiled into the binary, like `nav.js`.
pub const CAMERA_SCRIPT: &str = include_str!("camera.js");

/// Makes injection idempotent when a body is rewrapped.
const INJECTED_MARKER: &str = "__soli_camera_injected";

/// The attribute that turns the feature on for a page.
const CAMERA_ATTR: &str = "data-soli-camera";

fn camera_hash() -> u64 {
    use std::sync::OnceLock;
    static HASH: OnceLock<u64> = OnceLock::new();
    *HASH.get_or_init(|| prefetch::fnv1a_64(CAMERA_SCRIPT.as_bytes()))
}

/// Does this page use a camera? A page that does not gets no script.
pub fn page_uses_camera(html: &str) -> bool {
    html.contains(CAMERA_ATTR)
}

fn camera_tag() -> String {
    format!(
        "<!-- {} --><script src=\"/__soli/camera.js?v={:016x}\" defer></script>",
        INJECTED_MARKER,
        camera_hash()
    )
}

/// Insert the camera `<script>` before `</body>` — or `</html>`, or at the end.
pub fn inject_camera_tag(html: &str) -> String {
    if html.contains(INJECTED_MARKER) || !page_uses_camera(html) {
        return html.to_string();
    }
    let tag = camera_tag();
    if let Some(pos) = rfind_ascii_case_insensitive(html, b"</body>") {
        let mut out = String::with_capacity(html.len() + tag.len());
        out.push_str(&html[..pos]);
        out.push_str(&tag);
        out.push_str(&html[pos..]);
        out
    } else if let Some(pos) = rfind_ascii_case_insensitive(html, b"</html>") {
        let mut out = String::with_capacity(html.len() + tag.len());
        out.push_str(&html[..pos]);
        out.push_str(&tag);
        out.push_str(&html[pos..]);
        out
    } else {
        format!("{}{}", html, tag)
    }
}

/// `GET /__soli/camera.js`.
pub fn handle_camera_js() -> Response<Full<Bytes>> {
    Response::builder()
        .status(200)
        .header("Content-Type", "application/javascript; charset=utf-8")
        .header("Cache-Control", "public, max-age=86400, immutable")
        .body(Full::new(Bytes::from_static(CAMERA_SCRIPT.as_bytes())))
        .unwrap()
}

#[cfg(test)]
mod tests {
    use super::*;

    const WITH_CAMERA: &str = "<html><body><video data-soli-camera autoplay></video></body></html>";

    #[test]
    fn injects_for_a_page_with_a_camera() {
        let out = inject_camera_tag(WITH_CAMERA);
        assert!(out.contains("/__soli/camera.js?v="));
        assert!(out.find("camera.js").unwrap() < out.find("</body>").unwrap());
    }

    /// A page with no camera must not download a camera script.
    #[test]
    fn does_not_inject_otherwise() {
        let plain = "<html><body><p>hi</p></body></html>";
        assert_eq!(inject_camera_tag(plain), plain);
    }

    #[test]
    fn injection_is_idempotent() {
        let once = inject_camera_tag(WITH_CAMERA);
        assert_eq!(inject_camera_tag(&once), once);
    }

    /// Scanning implies a camera, so the scan attribute alone must not slip a
    /// page past the gate without the preview attribute that starts the stream.
    #[test]
    fn the_scan_attribute_travels_with_the_camera_attribute() {
        let scan_only = "<html><body><video data-soli-scan=\"qr_code\"></video></body></html>";
        assert_eq!(inject_camera_tag(scan_only), scan_only);
    }
}
